### PR TITLE
chore: add balls and bins simulator

### PR DESCRIPTION
**Pull request by romange**

### CODE REVIEW

File: tools/balls_bins.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    )
+
+    print(f"Probability that at least one bin has {args.high_threshold} or more balls: {atleast_p}")
+    if args.exact_num is not None:
+        print(f"Probability that at least one bin has {args.exact_num} balls: {exact_p}")
+
+    print(
+        f"Histogram of the difference between the most and least populated bins for {args.trials} trials"
+    )
+    plt.hist(deltas, bins=30, color="steelblue", edgecolor="none")
```
Comment (kostasrim) [inline tools/balls_bins.py:67]: Why are you limiting to 30 bins? I guess if we experiment with a large size of `args.bins` we will get large `deltas` and a lot of the data points will end up on the last bucket (bin 30). Is this intented?

**Review by kostasrim — APPROVED** [review]

File: tools/balls_bins.py (hunk 1 above)
Comment (romange) [reply tools/balls_bins.py:67]: 30 bins is 30 buckets that I assume would be defined according to the data distribution (auto bins).

File: tools/balls_bins.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python3
+
+"""Simulate throwing balls into bins."""
+
+import numpy as np
+import argparse
+import matplotlib.pyplot as plt
+
+
+def simulate_balls_into_bins(balls: int, N, threshold: int, exact, trials=10000):
```
[the remark this answers is not in the document]
Comment (romange) [reply tools/balls_bins.py]: fixed

File: tools/balls_bins.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    deltas = []
+
+    for _ in range(trials):
+        # Reset counts for each trial
+        counts.fill(0)
+
+        # Throw M balls into the bins
+        bins = np.random.randint(0, N, balls)
+        unique, counts_bins = np.unique(bins, return_counts=True)
+        counts[unique] += counts_bins
```
[the remark this answers is not in the document]
Comment (romange) [reply tools/balls_bins.py:24]: I asked chatgpt to write this code. It first wrote a correct code without numpy and I asked him to use numpy and he changed it to this. I would not know to write such code myself @kostasrim @dranikpg

**Pull request merged**